Lib/test/test_asyncgen.py

```diff
@@ -1728,6 +1754,111 @@ async def gen():
             gc_collect()
 
 
+class TestIssueGH74956(unittest.TestCase):
+    # simultanous use of generator by different coroutines is not
+    # allowed.
+    # https://github.com/python/cpython/issues/74956
+
+    def setUp(self):
+        self.loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(None)
+
+    def tearDown(self):
+        self.loop.close()
+        self.loop = None
+        asyncio.set_event_loop_policy(None)
+
+    def test_simultaneous_asend(self):
+        """
+        Verify that simultaneous use of generator by different coroutines is not
+        permitted.  We use Tasks to achieve this, where one task is suspended
+        in a `asyncio.sleep()` call inside the generator (during an `asend()` call),
+        and the other task attempts
+        to do an `asend()`, (or `athrow()`, or `aclose()`) on the generator.
+        """
+
+        
+        async def run_collision(op, *args):
+            # Two tasks are created and scheduled.  The first will sleep inside the
+            # `asend()` and the other will then attempt a second operation and fail.
+
+            async def consumer():
+                while True:
+                    # task fa will sleep here, and another task will try to iterate
+                    # the generator
+                    await asyncio.sleep(0)
+                    if (yield) is None:
+                        break
+
+            # create and start the generator
+            agenerator = consumer()
+            await agenerator.asend(None)
+
+            # start the first asend() task
+            fa = asyncio.create_task(agenerator.asend("A"))
+
+            # start the second task, which should fail (asend, athrow, aclose)
+            method = getattr(agenerator, op)
+            fb = asyncio.create_task(method(*args))
+
+            # first asend should succeed
+            await fa
+
+            # second operation should fail
+            with self.assertRaises(RuntimeError) as err:
+                await fb
+            assert "already running" in str(err.exception)
```

Comment:
I wonder if this order is well-defined.
I guess that asend is internally synchronous send + await, so maybe?

Reply:
With the current event loop, yes it is well defined. Tasks are scheduled in the order they are created.  First task runs until it hits `sleep()`, at which point the second task runs.  It is a strictly FIFO system.

asend is slightly more complex than send and await hooked together.  It needs to have its own "await" implementation.
If you are interested, my implementation of the async generator protocol using a my `Monitor` pattern is here:

https://github.com/kristjanvalur/py-asynkit/blob/f7929b4a5bc00391242189ace96cfc006862d9a0/src/asynkit/monitor.py#L337C13-L337C13